File: services/game/ai/physics.py
from constants import *

def predict_ball_trajectory(ball_pos, ball_v):
    """Simulate the path and collisions to get the Y where the ball should cross the goal 
    """
    current_pos = list(ball_pos)
    ball_v = list(ball_v)
    while (current_pos[0] + BALL_DIAM) <= PADDLE_AI_X and (current_pos[0] - BALL_DIAM) >= PADDLE_PLAYER_X:
        if (current_pos[1] - BALL_DIAM) < -(MAP_HEIGHT / 2) or (current_pos[1] + BALL_DIAM) > MAP_HEIGHT / 2:
            ball_v[1] = -ball_v[1]
        current_pos[1] += ball_v[1]
        current_pos[0] += ball_v[0]
    if (current_pos[0] > 0):
        current_pos[0] = min(PADDLE_AI_X - BALL_DIAM, current_pos[0])
    else:
        current_pos[0] = max(PADDLE_PLAYER_X + BALL_DIAM, current_pos[0])
    current_pos[1] = round(current_pos[1], 2)
    return current_pos

def collision_with_paddle(parent, child_paddle_y):
    """Use the parent predicted collision position
    return the velocity of the ball after the collision with the paddle
    """
    distance_center_paddle = child_paddle_y - PADDLE_HEIGHT + parent.future_ball_pos[1] if (abs(child_paddle_y - PADDLE_HEIGHT + parent.future_ball_pos[1])) else child_paddle_y + PADDLE_HEIGHT - parent.future_ball_pos[1]
    ball_v = [round(parent.ball_v[0] * -BALL_ACCELERATION, 2), round(parent.ball_v[1] * BALL_ACCELERATION * 0.16 * 1.05 * distance_center_paddle, 2)]
    return ball_v

File: services/game/ai/src/physics.js
import {
  MAP_HEIGHT,
  BALL_DIAM,
  PADDLE_PLAYER_X,
  PADDLE_AI_X,
  PADDLE_WIDTH
} from './constants.js';

/**
 * Predicts the Y-position of the ball when it reaches a paddle.
 * @param {x_b, y_b} ballPos. Current position of the ball  
 * @param {*} ballVel. Velocity of the ball.
 * @returns ballState of the ball when it reaches the paddle. v_by is switched depending on the bounces.
 */
export function predictBallState(ballPos, ballVel) {
  const [x_b, y_b] = ballPos;
  let [v_bx, v_by] = ballVel;

  const mapMax = (MAP_HEIGHT - BALL_DIAM) * 0.5;

  let timeToPaddle, nextReceiverX;

  if (v_bx > 0) {
    timeToPaddle = (Math.abs(PADDLE_PLAYER_X - x_b) - (PADDLE_WIDTH + BALL_DIAM) * 0.5) / (Math.abs(v_bx) + 0.0001);
    nextReceiverX = PADDLE_PLAYER_X - (PADDLE_WIDTH + BALL_DIAM) * 0.5;
  } else {
    timeToPaddle = (Math.abs(PADDLE_AI_X - x_b) - (PADDLE_WIDTH + BALL_DIAM) * 0.5) / (Math.abs(v_bx) + 0.0001);
    nextReceiverX = PADDLE_AI_X + (PADDLE_WIDTH + BALL_DIAM) * 0.5;
  }

  const predictedY = y_b + v_by * timeToPaddle;

  const playableHeight = MAP_HEIGHT - BALL_DIAM;
  let yInPlayable = predictedY + mapMax;

  let currentVy = v_by;
  if (yInPlayable < 0 || yInPlayable > playableHeight) {
    let workingY = yInPlayable;
    let bounceCount = 0;

    while (workingY < 0 || workingY > playableHeight) {
      if (workingY < 0) {
        workingY = -workingY;
        bounceCount++;
      } else if (workingY > playableHeight) {
        workingY = 2 * playableHeight - workingY;
        bounceCount++;
      }
    }

    yInPlayable = workingY;

    if (bounceCount % 2 === 1) {
      currentVy = -currentVy;
    }
  }

  const finalY = yInPlayable - mapMax;

  return {
    ballPos: [nextReceiverX, Math.round(finalY * 100) / 100],
    ballVel: [v_bx, v_by]
  };
}
